dll/win32/shell32/dialogs/folder_options.cpp

```diff
@@ -1388,12 +1392,71 @@ InsertFileType(HWND hDlgCtrl, WCHAR * szName, PINT iItem, WCHAR * szFile)
             RegQueryValueExW(hKey, L"EditFlags", NULL, NULL, (LPBYTE)&Entry->EditFlags, &dwSize);
     }
 
+    /* FIXME: Entry->AppName */
+
+    /* get icon */
+    HKEY hDefIconKey;
+    Entry->IconLocation[0] = UNICODE_NULL;
+    Entry->hIconLarge = Entry->hIconSmall = NULL;
+    if (RegOpenKeyExW(hKey, L"DefaultIcon", 0, KEY_READ, &hDefIconKey) == ERROR_SUCCESS)
+    {
+        WCHAR sz[MAX_PATH + 64];
+        dwSize = sizeof(sz);
+        if (!RegQueryValueExW(hDefIconKey, NULL, NULL, NULL, (LPBYTE)sz, &dwSize))
+        {
+            if (ExpandEnvironmentStringsW(sz, Entry->IconLocation,
+                                          _countof(Entry->IconLocation)))
+            {
+                LPWSTR pch = wcsrchr(Entry->IconLocation, L',');
+                if (pch)
+                {
+                    *pch = UNICODE_NULL;
+                    INT Index = _wtoi(pch + 1);
+                    if (Index < 0)
+                    {
+                        // A negative value will be interpreted as a resource ID.
+                        Index = -Index;
+                        HINSTANCE hDLL = LoadLibraryExW(Entry->IconLocation, NULL, LOAD_LIBRARY_AS_DATAFILE);
+                        Entry->hIconLarge = LoadIconW(hDLL, MAKEINTRESOURCEW(Index));
+                        Entry->hIconSmall = (HICON)LoadImageW(hDLL, MAKEINTRESOURCEW(Index), IMAGE_ICON,
+                                                              16, 16, 0);
+                        FreeLibrary(hDLL);
+                    }
+                    else
+                    {
+                        ExtractIconExW(Entry->IconLocation, Index,
+                                       &Entry->hIconLarge, &Entry->hIconSmall,
+                                       1);
+                    }
+                    *pch = L',';
+                }
+                else
+                {
+                    ExtractIconExW(Entry->IconLocation, 0,
+                                   &Entry->hIconLarge, &Entry->hIconSmall,
+                                   1);
+                }
+            }
+        }
+        RegCloseKey(hDefIconKey);
+    }
+
     /* close key */
     RegCloseKey(hKey);
 
+    INT iLargeImage = -1, iSmallImage = -1;
+    if (Entry->hIconLarge && Entry->hIconSmall)
+    {
+        iLargeImage = ImageList_AddIcon(himlLarge, Entry->hIconLarge);
+        iSmallImage = ImageList_AddIcon(himlSmall, Entry->hIconSmall);
+    }
+    assert(iLargeImage == iSmallImage);
```

Comment:
Shouldn't an `ASSERT` be upper-case?

Reply:
I have no idea that which is the better.

Reply:
Yeah, ASSERT is preferred. assert() comes from the CRT and will give you a message box, while ASSERT goes to the kernel debugger, so it's better for testbot etc.

Reply:
Done.

Reply:
Done.